fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -253,51 +254,64 @@ private boolean shouldClearExistingIndexEntries(@Nonnull IndexState indexState,
-                                if (bytes == null || fallbackMode) {
-                                    transaction.set(stampKey, stamp);
-                                } else if (! matchingTypeStamp(bytes)) {
-                                    if (LOGGER.isWarnEnabled()) {
-                                        LOGGER.warn(KeyValueLogMessage.of("Index was partly built by another method",
-                                                LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
-                                                LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
-                                                LogMessageKeys.INDEXER_ID, common.getUuid(),
-                                                LogMessageKeys.EXPECTED_TYPE, stamp,
-                                                LogMessageKeys.ACTUAL_TYPE, bytes));
-                                    }
-                                    throw new RecordCoreException("This index was partly built by another method");
-                                }
-                                return null;
-                            }); },
-                false,
-                additionalLogMessageKeyValues);
+                    .thenApply( bytes -> {
+                        if (bytes == null || forceStampOverwrite) {
+                            transaction.set(stampKey, indexingTypeStamp.toByteArray());
+                        } else {
+                            IndexBuildProto.IndexBuildIndexingStamp savedStamp ;
+                            try {
+                                savedStamp = IndexBuildProto.IndexBuildIndexingStamp.parseFrom(bytes);
+                            } catch (InvalidProtocolBufferException ex) {
+                                savedStamp = null;
+                            }
+                            if ( savedStamp == null || ! matchingIndexingTypeStamp(savedStamp)) {
+                                throw new RecordCoreException("This index was partly built by another method",
+                                        LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                                        LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
+                                        LogMessageKeys.INDEXER_ID, common.getUuid(),
+                                        LogMessageKeys.EXPECTED_TYPE, indexingTypeStamp,
+                                        LogMessageKeys.ACTUAL_TYPE, savedStamp);
+                            }
+                        }
+                        return null;
+                    });
+        } ));
     }
 
     @Nonnull
-    abstract byte[] getTypeStamp();
+    abstract IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp();
 
     @Nonnull
-    abstract boolean matchingTypeStamp(byte[] stamp);
+    abstract boolean matchingIndexingTypeStamp(@Nonnull final IndexBuildProto.IndexBuildIndexingStamp stamp);
```

Comment:
Is there a reason that this isn't `getIndexingTypeStamp().equals(stamp)`? I think that has the same semantics you've implemented, but is resilient to adding additional fields later.

It also probably isn't required for version 1, but having a more descriptive diff and therefore a more specific error message probably be good (some day).